crates/wasi-threads/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+
+pub struct WasiThreadsCtx<T> {
+    module: Module,
+    linker: Arc<Linker<T>>,
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Arc<Linker<T>>) -> Result<Self> {
+        if !has_wasi_entry_point(&module) {
+            bail!(
+                "failed to find wasi-threads entry point function: {}",
+                WASI_ENTRY_POINT
+            );
+        }
+        Ok(Self { module, linker })
+    }
+
+    pub fn spawn(&self, host: T, thread_start_arg: i32) -> Result<i32> {
+        let module = self.module.clone();
+        let linker = self.linker.clone();
+
+        // Start a Rust thread running a new instance of the current module.
+        let wasi_thread_id = random_thread_id();
+        let builder = thread::Builder::new().name(format!("wasi-thread-{}", wasi_thread_id));
+        builder.spawn(move || {
+            // Catch any panic failures in host code; e.g., if a WASI module
+            // were to crash, we want all threads to exit, not just this one.
+            let result = catch_unwind(AssertUnwindSafe(|| {
+                // Each new instance is created in its own store.
+                let mut store = Store::new(&module.engine(), host);
+
+                // Ideally, we would have already checked much earlier (e.g.,
+                // `new`) whether the module can be instantiated. Because
+                // `Linker::instantiate_pre` requires a `Store` and that is only
+                // available now. TODO:
+                // https://github.com/bytecodealliance/wasmtime/issues/5675.
+                let instance = linker.instantiate(&mut store, &module).expect(&format!(
+                    "wasi-thread-{} exited unsuccessfully: failed to instantiate",
+                    wasi_thread_id
+                ));
+                let thread_entry_point = instance
+                    .get_typed_func::<(i32, i32), ()>(&mut store, WASI_ENTRY_POINT)
+                    .unwrap();
+
+                // Start the thread's entry point. Any traps or calls to
+                // `proc_exit`, by specification, should end execution for all
+                // threads. This code uses `process::exit` to do so, which is what
+                // the user expects from the CLI but probably not in a Wasmtime
+                // embedding.
+                log::trace!(
+                    "spawned thread id = {}; calling start function `{}` with: {}",
+                    wasi_thread_id,
+                    WASI_ENTRY_POINT,
+                    thread_start_arg
+                );
+                match thread_entry_point.call(&mut store, (wasi_thread_id, thread_start_arg)) {
+                    Ok(_) => log::trace!("exiting thread id = {} normally", wasi_thread_id),
+                    Err(e) => {
+                        log::trace!("exiting thread id = {} due to error", wasi_thread_id);
+                        let e = maybe_exit_on_error(e);
```

Comment:
Terminating the parent process is a really heavy hammer and makes this whole crate unsuitable for use in multi-tenant embeddings. It would be helpful to understand exactly why this crate had to make that design choice, and to please note it somewhere high level, like in a crate-level README

Reply:
Hopefully this hammer is only temporary; @alexcrichton and I had discussed some early thoughts on how we might shut down all of the threads when necessary. It is complex to do so, though, so providing experimental wasi-threads access sooner for some users seemed more valuable than solving the problem and waiting longer.